### ISSUE TITLE
API RFC: Allow solutions callback in API

### ISSUE THREAD

**Issue by steelbrain** [steelbrain/linter#1705]
Currently `solutions` in Linter messages are an array, their callbacks can still be async tho. Nikolay pointed out in https://github.com/orgs/atom-ide-community/teams/atom-community/discussions/1/comments/5 that it would help with integrations in LSP-context if we made the solutions API async.

The usecase makes sense to me and it's not a major change. The API version will have to be bumped, semver-minor

**Comment by lierdakil** [comment]
To make it clear(er), here is roughly what I had in mind (TypeScript syntax because I don't grok flowjs):

```ts
type SolutionsArray = Array<
  | {
      title?: string
      position: Range
      priority?: number
      currentText?: string
      replaceWith: string
    }
  | {
      title?: string
      position: Range
      priority?: number
      apply: () => any
    }
>
type Message = {
  // ...
  solutions?: SolutionsArray | (() => Promise<SolutionsArray>)
  // ...
}
```

But I think it'd be best to first poke around and experiment with some approaches in https://github.com/atom-ide-community/atom-ide-languageclient to see if it actually makes sense before implementing it in linter API. I might find time to do that sometime this week.

**Comment by illright** [comment]
Yep, that makes sense in terms of the LSP API. The `textDocument/publishDiagnostics` notification will give us a list of diagnostics which we convert to messages, but in order to get the solutions for those messages, we must submit a `textDocument/codeAction` request, linking in the relevant diagnostics, which is exactly the job for this async callback.
